keys.cc

```diff
@@ -130,6 +130,16 @@ bound_kind invert_kind(bound_kind k) {
     abort();
 }
 
+bound_kind reverse_kind(bound_kind k) {
+    switch (k) {
+    case bound_kind::excl_start: return bound_kind::excl_end;
+    case bound_kind::incl_start: return bound_kind::incl_end;
+    case bound_kind::excl_end:   return bound_kind::excl_start;
+    case bound_kind::incl_end:   return bound_kind::incl_start;
+    }
+    abort();
```

Comment:
I don't like aborts and assertions (see https://github.com/scylladb/scylla/issues/7871) but if we must, can we at least have a message here, like an assertion failure message, so we can see the reason for the crash even before getting a full core?

Reply:
The compiler will warn if we forget an enum, so this is just to shut it up about needing to return something, and as a last resort.

Reply:
If the compiler thought that these enum values are the *only* possibilities, it wouldn't expect a return in this case. So if does, it means that it knows that although these are the only values the enum **should** hold, it may in fact hold any other integer value.

I agree that if we ever get to this abort() it's a bug - some other code mis-initialized an object or overwrote it it in memory. What I'm not sure about is why we can't kill (with an exception) the operation which relies on this broken data. Or, if we want an abort(), at least print a message so that the person who gets this crash will at least have a guess what went wrong before opening the core file.

Reply:
> If the compiler thought that these enum values are the _only_ possibilities, it wouldn't expect a return in this case. So if does, it means that it knows that although these are the only values the enum **should** hold, it may in fact hold any other integer value.
> 

No, that's undefined behavior (at least for regular enum; it's defined for `enum class : base_type`.

> I agree that if we ever get to this abort() it's a bug - some other code mis-initialized an object or overwrote it it in memory. What I'm not sure about is why we can't kill (with an exception) the operation which relies on this broken data. Or, if we want an abort(), at least print a message so that the person who gets this crash will at least have a guess what went wrong before opening the core file.

There is on_internal_error().